### PR TITLE
Issue #1020: Dynamically load second-step visibility conditions confi…

### PR DESCRIPTION
**Pull request by docwilmot**
…guration in the same dialog

Fixes https://github.com/backdrop/backdrop-issues/issues/1020
Fixes https://github.com/backdrop/backdrop-issues/issues/943
Fixes https://github.com/backdrop/backdrop-issues/issues/1022

### CODE REVIEW

File: core/modules/layout/layout.module [hunk 1]
```diff
@@ -242,14 +223,6 @@ function layout_menu() {
...
   ) + $base;
 
-  $items['admin/structure/layouts/menu/%layout_tempstore_menu_item/condition/edit/%'] = array(
-    'title' => 'Configure condition',
-    'title callback' => 'layout_condition_edit_title',
-    'title arguments' => array(NULL, NULL, 4, 7),
-    'page callback' => 'backdrop_get_form',
-    'page arguments' => array('layout_condition_edit_form', NULL, NULL, 4, 7),
-    'type' => MENU_CALLBACK,
-  ) + $base;
```
Comment (quicksketch) [inline core/modules/layout/layout.module:252]: As noted in https://github.com/backdrop/backdrop-issues/issues/1020#issuecomment-186708336, removal of these callbacks breaks the links to Configure an existing condition after it's been added. We may need to keep these callbacks and the matching forms.

File: core/modules/layout/layout.admin.inc [hunk 2]
```diff
@@ -1498,54 +1531,35 @@ function layout_condition_edit_form($form, &$form_state, Layout $layout = NULL,
   $form_state['layout'] = $layout;
   $form_state['block'] = $block;
   $form_state['menu_item'] = $menu_item;
-  $form_state['handler'] = $handler;
+  $form_state['handler']  = $handler;
```
Comment (klonos) [inline core/modules/layout/layout.admin.inc:1583]: extraneous space creeped there?

**Comment by klonos** [comment]
I was playing around with this PR (-the changes that @quicksketch said should be left in place) when I got this while in the condition add dialog:

```
Notice: Undefined index: layout_condition_edit_form in backdrop_retrieve_form() (line 806 of C:\www\backdrop\core\includes\form.inc).
Warning: call_user_func_array() expects parameter 1 to be a valid callback, function 'layout_condition_edit_form' not found or invalid function name in backdrop_retrieve_form() (line 850 of C:\www\backdrop\core\includes\form.inc).
Notice: Undefined index: layout_condition_edit_form in backdrop_retrieve_form() (line 806 of C:\www\backdrop\core\includes\form.inc).
Warning: call_user_func_array() expects parameter 1 to be a valid callback, function 'layout_condition_edit_form' not found or invalid function name in backdrop_retrieve_form() (line 850 of C:\www\backdrop\core\includes\form.inc).
```

...I think it was after I added a condition and then removed it without submitting the config dialog and then trying to re-add a new condition (still not having closed the dialog).

**Comment by klonos** [comment]
...I was trying to reproduce this, and I got an empty config condition dialog:

backdrop-issue1020-pull1260-empty_condition_config_dialog_when_condition_not_saved

To reproduce:
1. Configure a block and add a condition - DO NOT SAVE/CLOSE THE DIALOG.
2. Hit the "Configure" link of the newly added condition -> blank dialog
3. Close the empty dialog.
4. Hit the "Configure" link of the block you've been editing before. -> the condition is there although you have not saved config.
5. [optional] Remove the condition you previously added (BUT HAVE NOT SAVED). It still leads to an empty dialog and cannot be configured anyways.
6. Add another condition -> soon as you make any selection from the "Add visibility condition" drop-down menu, you get the warning.

**Comment by docwilmot** [comment]
Thanks, I have a start on fixing the above. Will push sometime before 1.4.0 :smile:

**Comment by klonos** [comment]
:smile:

**Comment by docwilmot** [comment]
Fixed. Added test for clicking the configure link.

File: core/modules/layout/layout.admin.inc [hunk 3]
```diff
@@ -18,7 +18,7 @@ function layout_list_page() {
     $path_groups[$layout->getPath()][$layout->name] = $layout;
   }
 
-  // Assmeble the rows of the table.
+  // Assemble the rows of the table.
```
Comment (klonos) [inline core/modules/layout/layout.admin.inc:21]: Good catch :+1: :smile:

File: core/modules/layout/layout.admin.inc (hunk 2 above)
Comment (klonos) [inline core/modules/layout/layout.admin.inc:1583]: ???

File: core/modules/layout/layout.module [hunk 4]
```diff
@@ -244,13 +240,12 @@ function layout_menu() {
...
     'title' => 'Configure condition',
-    'title callback' => 'layout_condition_edit_title',
-    'title arguments' => array(NULL, NULL, 4, 7),
     'page callback' => 'backdrop_get_form',
-    'page arguments' => array('layout_condition_edit_form', NULL, NULL, 4, 7),
+    'page arguments' => array('layout_condition_add_form', NULL, NULL, 4, 7),
     'type' => MENU_CALLBACK,
   ) + $base;
 
+
```
Comment (klonos) [inline core/modules/layout/layout.module:248]: Why the extra line here??

**Comment by klonos** [comment]
Much better this time. Still a few UX nits here and there (same steps as above more or less):
1. Configure a block and add a condition - DO NOT SAVE/CLOSE THE DIALOG.
2. Hit the "Configure" link of the newly added condition -> no blank dialog this time, **but** the submit button should say "**Save** visibility condition configuration" not "**Add** visibility condition" (we can either omit the "configuration" part if you think it is too verbose or just use "**Save** configuration" instead).
3. Close the empty dialog using the x button at the top right - NOT the "Save configuration" button. -> The "dirty form" message is there on the page although you did not save block configuration. WTF?
4. Hit the "Configure" link of the block you've been editing before. -> the condition is there although you have not saved config. It is not shown in the block overview, but it is visible in the block configure dialog. WTF?
5. Remove the condition you previously added (BUT HAVE NOT SAVED). This time do save block configuration. -> WTF is the "dirty form" message doing there? I added a visibility condition to a block that I did not save. When I edited the block the visibility condition was there. I removed it **and** saved config this time, so I am back to where I started **without** any actual changes. I repeat: WTF is the "dirty form" message doing there?

...guess I'm kicking the tires a bit too much here :smile:

**Comment by quicksketch** [comment]
The PR has been updated to fix the issue noted above. I found a new problem that I wrote up in the original issue: https://github.com/backdrop/backdrop-issues/issues/1020#issuecomment-200000038

**Comment by docwilmot** [comment]
Merged in the PR to my branch from @quicksketch so if Travis likes this, it should be ready for a final review and RTBC.

**Comment by klonos** [comment]
Not a show-stopper or an annoyance. Doesn't bother me that much, but kinda looks silly:
1. add new layout
2. give it a name
3. select template
4. define a path
5. add a condition -> you get the "_This form has unsaved changes. Click "Save layout" to make changes permanent or "Cancel" to discard changes._" message soon as you add the condition.

There are no changes per se because this layout is a new one to be added. Either show this message from step 1 or suppress it if the layout is not an existing one.

**Comment by klonos** [comment]
...also (in this case of a new layout), we prompt the user through the message to "_Click "Save layout"_" while the submit button is "Create layout".

**Comment by klonos** [comment]
...visibility condition "Is the front page" needs a bit more love. Perhaps "Current page is the front page" or something.

**Comment by klonos** [comment]
...visibility condition "Node is "Page"" should perhaps be "Content type is "Page"" or "Content is of type "Page"".

**Comment by klonos** [comment]
...visibility condition "User has "xyz" permission." perhaps with an added "the": "User has **the** "xyz" permission."

**Comment by klonos** [comment]
...visibility condition "User has the role Administrator" perhaps wrap the role in double quotes for consistency like we do with the content type and the permission: "User has the role "Administrator"".

Similarly when multiple roles are selected:

User has one of the following roles: "Authenticated", "Administrator"

instead of:

User has one of the following roles: Authenticated, Administrator

**Comment by klonos** [comment]
...still some of the behavior I described earlier (minus the things that have been fixed):
1. Edit an existing layout.
2. Hit the "Configure" button on any of the existing blocks (if you use the header block though, it'd be easier to see the message).
3. Click the "add visibility" link.
4. Add any condition you want and hit the "Add visibility condition" button -> notice the "dirt form" message.
5. DO NOT CLICK the "Save configuration" button. Close the dialog using the x on the top-right corner -> the "dirty form" message is there although the configuration dialog was not saved.

Another example:
1. Edit an existing layout.
2. Hit the "Configure" button on any of the existing blocks (if you use the header block though, it'd be easier to see the message).
3. Click the "add visibility" link.
4. Add any condition you want and hit the "Add visibility condition" button.
5. This time DO click the "Save configuration" button.
6. Edit another block and add a visibility to that one too and save configuration.
7. Finally, save the layout changes.
8. Re-edit the same layout.
9. Edit some more blocks and add conditions or change other settings.
10. Edit one of the blocks with the added conditions that you have edited the first time and either remove or change the condition you have previously added.
11. Changed your mind (perhaps realized it was the other block you actually wanted to edit)... hm, two options:
    - save the layout and **try to remember** all the changes that were made to the last block so you can re-edit the layout (oh the number of clicks!) and revert them. Or...
    - Hit the "Cancel" button on the layout form so you can discard the unintentional changes made when editing the last block (step 10) AND along with them, **lose all the intentional changes you made in step 9**.

I think that a "Cancel" button is in order for the block configuration dialog. If the user hits it, whatever changes were made to the config **of the specific block** (including any added/changed/removed conditions) should be discarded. This way, any changes to the configuration of other blocks made previously during the same edit session of the same layout are not lost.

**Comment by quicksketch** [comment]
I tested this out again and I think it's ready to go. I think @klonos's findings are valid though, but none of the issues mentioned are introduced by this change. This solves 3 existing problems, so I believe we should merge this as-is and then create another follow-up for all the new findings.

**Comment by klonos** [comment]
Feel free to merge away and I'll re-test and file follow-ups as needed :wink:

**Comment by quicksketch** [comment]
Great, merged into 1.x for 1.4.0. I'll fix that extra space problem pointed out as well.

**Comment by klonos** [comment]
Thanx. I will go through the list of comments above and re-test to see if everything is sorted. If not, I'll file follow-up issues as needed.

**Pull request merged**